Fix parsing and precision handling for easting and northing values

mdang8 opened the pull request:
## Changes

- Regex for parsing easting and northing values of MGRS coordinates to be able to split them into equal length strings
- Precision determination to handle easting and northing values containing leading zeros

Fixes #17 
@brendan-hofmann @mcalcote @figliold @josephthweatt @dcruver @dimitry-sherman

mdang8 committed:
Fix parsing and precision handling for easting and northing values

mdang8 committed:
Fix formatting issues

brendan-hofmann commented on usng4j-api/src/main/java/org/codice/usng4j/UsngCoordinate.java:
```diff
@@ -59,7 +59,7 @@
 
   /** RegEx expressions for MGRS Northing and Easting parsing */
-  public final String MGRS_COORDINATE_PART_REGEX_STRING = "(\\d{0,5})\\W*(\\d{0,5})\\W*";
+  public final String MGRS_COORDINATE_PART_REGEX_STRING = "(\\d{0,12})\\W*";
```
What non-word characters are we hoping to match here?

mdang8 replied:
There shouldn't be any other characters to match after the numerical location. I changed the regex to only look for up to 10 digits.

brendan-hofmann commented on usng4j-api/src/main/java/org/codice/usng4j/CoordinatePrecision.java:
```diff
@@ -64,16 +64,16 @@ public int getIntValue() {
-  public static CoordinatePrecision forEastNorth(int easting, int northing) {
-    int maxValue = Math.max(easting, northing);
+  public static CoordinatePrecision forEastNorth(String easting, String northing) {
+    int maxLength = Math.max(easting.length(), northing.length());
```
Do we need to worry about trimming?

mdang8 replied:
We shouldn't since all of the easting/northing values should be trimmed already before calling `forEastNorth()` in the parse methods, but I added a trim just in case it gets missed in the future.

mdang8 committed:
Address PR comments

Change MGRS regex to only parse 10-digit max numerical locations and trim easting/northing strings passed to precision determination

josephthweatt commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/UsngCoordinateImpl.java:
```diff
@@ -106,7 +106,31 @@
+    this(zoneNumber, latitudeBandLetter, columnLetter, rowLetter);
+    this.easting = easting;
+    this.northing = northing;
+    this.precision = precision;
```
are these lines formatted?

mdang8 replied:
Yes, that's from running the maven formatter.

josephthweatt commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/UsngCoordinateImpl.java:
```diff
@@ -183,20 +207,44 @@ private static UsngCoordinate parseCoordinateString(
-      if (!(StringUtils.isEmpty(m.group(4)) || StringUtils.isEmpty(m.group(5)))) {
-        int easting = Integer.parseInt(m.group(4).trim());
-        int northing = Integer.parseInt(m.group(5).trim());
+      if (captureGroupsCount > 4
```
There's lot of 4's and 5's and magic numbers. Would it be possible to store these as constant int's to make the code more readable?

mdang8 replied:
I removed some of the code referencing those numbers and added some comments explaining what they mean.

josephthweatt reviewed (commented):
Leaving a couple questions.

figliold commented on usng4j-api/src/main/java/org/codice/usng4j/CoordinatePrecision.java:
```diff
@@ -64,16 +64,16 @@ public int getIntValue() {
 
-  public static CoordinatePrecision forEastNorth(int easting, int northing) {
-    int maxValue = Math.max(easting, northing);
+  public static CoordinatePrecision forEastNorth(String easting, String northing) {
```
Would you really be looking for the max length of easting and northing?  I believe for an MGRS to be valid both easting and northing have to be the same length.

figliold commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/UsngCoordinateImpl.java:
```diff
@@ -183,20 +207,44 @@ private static UsngCoordinate parseCoordinateString(
+                zoneNumber, latitudeBandLetter, columnLetter, rowLetter, eastingInt, northingInt);
+      } else if (!StringUtils.isEmpty(m.group(4))) {
+        // full numerical location with easting and northing values given as n+n digits
+        String numericalLocation = m.group(4).trim();
```
If easting length + northing length is not even, would we want to divide easting length + northing length by 2?  It seems like that would be an invalid MGRS.

mdang8 replied:
No we wouldn't want to use those easting/northing values because they would be invalid. I chose to add a check that will throw a `ParseException` if the numerical location can't be evenly split.

figliold commented on usng4j-impl/src/main/java/org/codice/usng4j/impl/UsngCoordinateImpl.java:
```diff
@@ -183,20 +207,44 @@ private static UsngCoordinate parseCoordinateString(
+        int splitIndex = numericalLocation.length() / 2;
+        // easting value is first n digits and northing value is last n digits
+        String easting = numericalLocation.substring(0, splitIndex);
+        String northing = numericalLocation.substring(splitIndex);
```
Wouldn't Integer.parseInt() drop leading zeros?  Does that cause any issues?

mdang8 replied:
That would cause an issue. I made a change to determine the precision of the coordinate before calling `Integer.parseInt()`. This way the precision should be retained throughout the rest of the parsing/conversions and the other math being done on the easting/northing integer values can be left unchanged.

figliold commented on usng4j-api/src/main/java/org/codice/usng4j/CoordinatePrecision.java:
```diff
@@ -64,16 +64,16 @@ public int getIntValue() {
+    int maxLength = Math.max(easting.trim().length(), northing.trim().length());
 
-    if (maxValue > 9999) {
+    if (maxLength > 4) {
```
Would this be something like:
if(easting.trim().length() == 5 && northing.trim().length() == 5) {
     return CoordinatePrecision.ONE_METER;
}
Something similar for the rest of these precision checks.

mdang8 replied:
I'm a bit unsure of how to handle when easting/northing values of unequal lengths get passed in to this method. There are tests where it looks like invalid coordinates are being used (e.g. https://github.com/codice/usng4j/blob/7ea805a0f5e26b279c46fdc2383d5f095d8b41de/usng4j-impl/src/test/java/org/codice/usng4j/impl/CoordinateSystemTranslatorTest.java#L287-L288) that use the `CoordinatePrecision.forEastNorth()` method and are expected to pass. I think it makes sense to throw an exception when easting/northing values of unequal lengths are used, which I did inside the coordinate parsing methods, but there's not currently any checks inside the `CoordinatePrecision.forEastNorth()` method because that would cause those existing tests to fail. I'm not sure if there's something else I'm overlooking, but if there isn't, then we could change those tests to use valid easting/northing values and add the checks.

mdang8 committed:
Fix USNG parsing to correctly split numerical location

mdang8 committed:
Fix invalid northing value used in test

mcalcote commented:
build now

figliold commented:
Hero'd successfully.

These are the steps I used to hero this PR:
1) Pull down this PR and run mvn clean install on codice/usng4j
2) Clone the codice/imaging-nitf repository and change the version
   in https://github.com/codice/imaging-nitf/blob/master/core/pom.xml 
   from 0.1 to 0.5-SNAPSHOT (the current locally built version of usng4j with this PR's changeset).
    <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <usng4j.version>0.5-SNAPSHOT</usng4j.version>
    </properties>
3) Run mvn clean install on codice/imaging-nitf
4) Verify the build is successful 

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Codice Imaging: NITF Parent ........................ SUCCESS [  2.064 s]
[INFO] Codice Imaging: NITF Shared Test Resources ......... SUCCESS [01:23 min]
[INFO] Codice Imaging: NITF Core API ...................... SUCCESS [01:20 min]
[INFO] Codice Imaging: NITF Core .......................... SUCCESS [07:48 min]
[INFO] Codice Imaging: NITF Metadata Comparison (Engineering Tool) SUCCESS [  2.706 s]
[INFO] Codice Imaging: CGM ................................ SUCCESS [01:12 min]
[INFO] Codice Imaging: Image Compare ...................... SUCCESS [01:45 min]
[INFO] Codice Imaging: NITF Render ........................ SUCCESS [10:08 min]
[INFO] Codice Imaging: NITF Fluent API .................... SUCCESS [ 19.433 s]
[INFO] Codice Imaging: NITF Fluent ........................ SUCCESS [ 59.095 s]
[INFO] Codice Imaging: TRE Wrappers ....................... SUCCESS [ 47.383 s]
[INFO] Codice Imaging: NITF Registry Values Parser ........ SUCCESS [  4.573 s]
[INFO] Codice Imaging: DES Wrappers ....................... SUCCESS [ 37.795 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 26:36 min
[INFO] Finished at: 2019-03-14T10:30:56-07:00
[INFO] Final Memory: 77M/1608M
[INFO] ------------------------------------------------------------------------
